File: clients/node/src/sidecar.js
'use strict';

const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const https = require('node:https');
const crypto = require('node:crypto');

const BINARY = 'amazon-dynamodb-streams-consumer-sidecar';
const VERSION = '0.1.0';
const RELEASE_BASE =
  process.env.DDB_STREAMS_CONSUMER_RELEASE_BASE ||
  'https://github.com/LeeroyHannigan/amazon-dynamodb-streams-consumer/releases/download';

// Maps Node's platform/arch to the stable release asset naming (os in
// {linux,darwin,windows}, arch in {x86_64,aarch64}) -- the same language-neutral
// contract release.yml publishes and the Go client consumes.
function platformArch() {
  const osName =
    process.platform === 'win32' ? 'windows' : process.platform === 'darwin' ? 'darwin' : 'linux';
  const arch = process.arch === 'x64' ? 'x86_64' : process.arch === 'arm64' ? 'aarch64' : process.arch;
  const ext = osName === 'windows' ? '.exe' : '';
  return { osName, arch, ext };
}

function cachePath() {
  const base = process.env.XDG_CACHE_HOME || path.join(os.homedir(), '.cache');
  const { ext } = platformArch();
  return path.join(base, 'amazon-dynamodb-streams-consumer', VERSION, BINARY + ext);
}

// GET with redirect following (GitHub release assets 302 to a CDN).
function httpGet(url, redirects = 0) {
  return new Promise((resolve, reject) => {
    if (redirects > 5) return reject(new Error('too many redirects'));
    https
      .get(url, (res) => {
        const { statusCode, headers } = res;
        if (statusCode >= 300 && statusCode < 400 && headers.location) {
          res.resume();
          resolve(httpGet(headers.location, redirects + 1));
          return;
        }
        if (statusCode !== 200) {
          res.resume();
          reject(new Error(`GET ${url}: HTTP ${statusCode}`));
          return;
        }
        const chunks = [];
        res.on('data', (c) => chunks.push(c));
        res.on('end', () => resolve(Buffer.concat(chunks)));
        res.on('error', reject);
      })
      .on('error', reject);
  });
}

async function download(dst) {
  const { osName, arch, ext } = platformArch();
  const asset = `${BINARY}-${osName}-${arch}${ext}`;
  const binURL = `${RELEASE_BASE.replace(/\/$/, '')}/v${VERSION}/${asset}`;
  const want = (await httpGet(binURL + '.sha256')).toString().trim().split(/\s+/)[0];
  const body = await httpGet(binURL);
  const got = crypto.createHash('sha256').update(body).digest('hex');
  if (got.toLowerCase() !== want.toLowerCase()) {
    throw new Error(`checksum mismatch for ${asset}: got ${got} want ${want}`);
  }
  fs.mkdirSync(path.dirname(dst), { recursive: true });
  const tmp = `${dst}.tmp-${process.pid}`;
  fs.writeFileSync(tmp, body, { mode: 0o755 });
  fs.renameSync(tmp, dst);
  return dst;
}

function onPath() {
  const { ext } = platformArch();
  const name = BINARY + ext;
  for (const dir of (process.env.PATH || '').split(path.delimiter)) {
    if (!dir) continue;
    const p = path.join(dir, name);
    try {
      fs.accessSync(p, fs.constants.X_OK);
      return p;
    } catch {
      /* not here */
    }
  }
  return null;
}

// Resolution order: explicit path -> env override -> cached download -> download
// -> PATH. Unlike npm, we don't ship a per-platform binary in the tarball; the
// sidecar is fetched once and cached, so it is still install-and-go.
async function discoverSidecar(explicit) {
  if (explicit) return explicit;
  if (process.env.DDB_STREAMS_CONSUMER_SIDECAR) return process.env.DDB_STREAMS_CONSUMER_SIDECAR;
  const cached = cachePath();
  try {
    fs.accessSync(cached, fs.constants.X_OK);
    return cached;
  } catch {
    /* need to fetch */
  }
  try {
    return await download(cached);
  } catch (e) {
    const p = onPath();
    if (p) return p;
    throw new Error(
      `could not obtain the ${BINARY} sidecar: download failed (${e.message}) and it is not on PATH. ` +
        'Set DDB_STREAMS_CONSUMER_SIDECAR=/path/to/sidecar or install it manually'
    );
  }
}

module.exports = { discoverSidecar, cachePath, platformArch, VERSION };

File: clients/go/sidecar.go
package ddbstreams

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

const binaryName = "amazon-dynamodb-streams-consumer-sidecar"

// releaseBaseURL is where per-platform sidecar assets are published. Overridable
// via DDB_STREAMS_CONSUMER_RELEASE_BASE for testing / mirrors.
const releaseBaseURL = "https://github.com/LeeroyHannigan/amazon-dynamodb-streams-consumer/releases/download"

// discoverSidecar resolves the sidecar binary, in order:
//  1. explicit path argument,
//  2. DDB_STREAMS_CONSUMER_SIDECAR env override,
//  3. cached binary from a previous download,
//  4. transparent download from the matching GitHub Release (checksum-verified),
//  5. the binary on PATH.
//
// Unlike Python (whose wheel bundles the binary), Go fetches source, so the
// binary is materialized on first use and cached — from the user's seat it is
// still "import and go".
func discoverSidecar(explicit string) (string, error) {
	if explicit != "" {
		return explicit, nil
	}
	if env := os.Getenv("DDB_STREAMS_CONSUMER_SIDECAR"); env != "" {
		return env, nil
	}

	cached := cachePath()
	if fi, err := os.Stat(cached); err == nil && !fi.IsDir() {
		return cached, nil
	}

	if path, err := downloadSidecar(cached); err == nil {
		return path, nil
	} else if p, perr := lookPath(); perr == nil {
		return p, nil // fall back to PATH if download failed
	} else {
		return "", fmt.Errorf(
			"could not obtain the %q sidecar: download failed (%v) and it is not on PATH. "+
				"Set DDB_STREAMS_CONSUMER_SIDECAR=/path/to/sidecar or install it manually",
			binaryName, err)
	}
}

func lookPath() (string, error) {
	// Minimal PATH lookup without importing os/exec's LookPath semantics here.
	name := binaryName
	if runtime.GOOS == "windows" {
		name += ".exe"
	}
	for _, dir := range filepath.SplitList(os.Getenv("PATH")) {
		if dir == "" {
			continue
		}
		p := filepath.Join(dir, name)
		if fi, err := os.Stat(p); err == nil && !fi.IsDir() {
			return p, nil
		}
	}
	return "", fmt.Errorf("%s not found on PATH", name)
}

// platformArch maps Go's GOOS/GOARCH to the release asset naming
// (x86_64 / aarch64), matching the wheel platform tags.
func platformArch() (goos, arch, ext string) {
	goos = runtime.GOOS
	switch runtime.GOARCH {
	case "amd64":
		arch = "x86_64"
	case "arm64":
		arch = "aarch64"
	default:
		arch = runtime.GOARCH
	}
	if goos == "windows" {
		ext = ".exe"
	}
	return goos, arch, ext
}

func cachePath() string {
	base, err := os.UserCacheDir()
	if err != nil || base == "" {
		base = filepath.Join(os.TempDir(), "ddb-streams-consumer-cache")
	}
	_, _, ext := platformArch()
	return filepath.Join(base, "amazon-dynamodb-streams-consumer", Version, binaryName+ext)
}

func assetURL() (bin, sha string) {
	base := os.Getenv("DDB_STREAMS_CONSUMER_RELEASE_BASE")
	if base == "" {
		base = releaseBaseURL
	}
	goos, arch, ext := platformArch()
	asset := fmt.Sprintf("%s-%s-%s%s", binaryName, goos, arch, ext)
	u := fmt.Sprintf("%s/v%s/%s", strings.TrimRight(base, "/"), Version, asset)
	return u, u + ".sha256"
}

// downloadSidecar fetches the platform sidecar to dst, verifying its SHA-256
// against the published .sha256, and marks it executable.
func downloadSidecar(dst string) (string, error) {
	binURL, shaURL := assetURL()

	wantSum, err := fetchString(shaURL)
	if err != nil {
		return "", fmt.Errorf("fetch checksum %s: %w", shaURL, err)
	}
	wantSum = strings.TrimSpace(strings.Fields(wantSum)[0])

	body, err := fetchBytes(binURL)
	if err != nil {
		return "", fmt.Errorf("fetch sidecar %s: %w", binURL, err)
	}
	h := sha256.Sum256(body)
	gotSum := hex.EncodeToString(h[:])
	if !strings.EqualFold(gotSum, wantSum) {
		return "", fmt.Errorf("checksum mismatch for %s: got %s want %s", binURL, gotSum, wantSum)
	}

	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return "", err
	}
	// Write atomically: temp file in the same dir, then rename.
	tmp, err := os.CreateTemp(filepath.Dir(dst), ".sidecar-*")
	if err != nil {
		return "", err
	}
	tmpName := tmp.Name()
	defer os.Remove(tmpName)
	if _, err := tmp.Write(body); err != nil {
		tmp.Close()
		return "", err
	}
	if err := tmp.Chmod(0o755); err != nil {
		tmp.Close()
		return "", err
	}
	if err := tmp.Close(); err != nil {
		return "", err
	}
	if err := os.Rename(tmpName, dst); err != nil {
		return "", err
	}
	return dst, nil
}

func fetchBytes(url string) ([]byte, error) {
	resp, err := http.Get(url) //nolint:gosec // release URL, checksum-verified by caller
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GET %s: HTTP %d", url, resp.StatusCode)
	}
	return io.ReadAll(resp.Body)
}

func fetchString(url string) (string, error) {
	b, err := fetchBytes(url)
	return string(b), err
}
